log/loki.go

```diff
@@ -0,0 +1,486 @@
+					t1, t2, t3, t4, cutOffIndex, size)
+			}
+		}
+	}()
+
+	for {
+		select {
+		case entry := <-h.ch:
+			if count == h.limit {
+				dropped++
+				continue
+			}
+
+			// Arguably we can directly generate the final marshalled version of the labels right here
+			// through sorting the entry.Data, removing additionalparams from it and then dumping it
+			// as the final marshal and appending level and h.labels after it.
+			// If we reuse some kind of big enough `[]byte` buffer we can also possibly skip on some
+			// of allocation. Combined with the cutoff part and directly pushing in the final data
+			// type this can be really a lot faster and to use a lot less memory
+			labels := make(map[string]string, len(entry.Data)+1)
+			for k, v := range entry.Data {
+				labels[k] = fmt.Sprint(v) // TODO optimize ?
+			}
+			for _, params := range h.labels {
+				labels[params[0]] = params[1]
+			}
+			labels["level"] = entry.Level.String()
+			// have the cutoff here ?
+			// if we cutoff here we can cut somewhat on the backbuffers and optimize the inserting
+			// in/creating of the final Streams that we push
+			msgs[count] = tmpMsg{
+				labels: labels,
+				msg:    entry.Message,
+				t:      entry.Time.UnixNano(),
+			}
+			count++
+		case <-ticker.C:
+			ch := make(chan time.Duration)
+			pushCh <- ch
+			ch <- h.pushPeriod / 2
+			<-ch
+		case <-h.ctx.Done():
+			ch := make(chan time.Duration)
+			pushCh <- ch
+			ch <- 0
+			<-ch
+			return
+		}
+	}
+}
+
+func sortAndSplitMsgs(msgs []tmpMsg, bufferTime time.Duration) int {
+	if len(msgs) == 0 {
+		return 0
+	}
+
+	// TODO using time.Before was giving a lot of out of order, but even now, there are some, if the
+	// limit is big enough ...
+	sort.Slice(msgs, func(i, j int) bool {
+		return msgs[i].t < msgs[j].t
+	})
+
+	// We can technically cutoff during the addMsg phase, which will be even better if we sort after
+	// it as well ... maybe
+	cutOff := time.Now().Add(-bufferTime).UnixNano() // probably better to be configurable
+
+	cutOffIndex := sort.Search(len(msgs), func(i int) bool {
+		return !(msgs[i].t < cutOff)
+	})
+	return cutOffIndex
+}
+
+func (h *lokiHook) createPushMessage(msgs []tmpMsg, cutOffIndex, dropped int) *lokiPushMessage {
+	strms := new(lokiPushMessage)
+	strms.msgMaxSize = h.msgMaxSize
+	for _, msg := range msgs[:cutOffIndex] {
+		strms.add(msg)
+	}
+	if dropped != 0 {
+		labels := make(map[string]string, 2+len(h.labels))
+		labels["level"] = logrus.WarnLevel.String()
+		labels["dropped"] = strconv.Itoa(dropped)
+		for _, params := range h.labels {
+			labels[params[0]] = params[1]
+		}
+
+		msg := tmpMsg{
+			labels: labels,
+			msg: fmt.Sprintf("k6 dropped some packages because they were above the limit of %d/%s",
+				h.limit, h.pushPeriod),
+			t: msgs[cutOffIndex-1].t,
+		}
+		strms.add(msg)
+	}
+	return strms
+}
+
+func (h *lokiHook) push(b bytes.Buffer) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
```

Comment:
Can you add the timeout in the `lokiHook` struct, even if you don't make it configurable for now? It's very hidden here, and, to be honest, 5 seconds seem a bit too short when you could have large HTTP requests...

Reply:
This actually doesn't matter as the real timeout is set in the client ... I should probably just use `context.Background()` directly ?

Reply:
Or... don't use `http.NewRequestWithContext()`, simply use `http.NewRequest()` :stuck_out_tongue_closed_eyes:

Reply:
THen I need to `//nolint:noctx` or w/e it was ...